server/src/main/java/org/opensearch/transport/ProtocolInboundMessage.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.transport;
+
+import org.opensearch.common.annotation.ExperimentalApi;
+
+/**
+ * Base class for inbound data as a message.
+ * Different implementations are used for different protocols.
+ *
+ * @opensearch.internal
+ */
+@ExperimentalApi
+public interface ProtocolInboundMessage {
+
+    /**
+     * The protocol used to encode this message
+     */
+    static String NATIVE_PROTOCOL = "native";
+    static String PROTOBUF_PROTOCOL = "protobuf";
+
+    /**
+     * @return the protocol used to encode this message
+     */
+    default public String getProtocol() {
+        return NATIVE_PROTOCOL;
+    }
```

Comment:
I'm not sure I understand the value of adding these here, because it seems to limit future flexibility. It feels a little like a base class defining its subclasses.

Could this interface declare `getProtocol()`, but leave the implementation up to implementers? Then `NativeInboundMessage` could return define and return `NATIVE_PROTOCOL`. Then  `protocolMessageHandlers` could map from `NativeInboundMessage.PROTOCOL` to `NativeMessageHandler`, and later from `ProtobufInboundMessage.PROTOCOL` to `ProtobufMessageHandler`.

In some magical future, a plugin could inject its own mapping from a protocol string to a `ProtocolMessageHandler`.

Reply:
We added a default implementation for since it was changing `InboundMessage`. I think we can now move away from default implementation since we have a`NativeInboundMessage`. Let me try to work this in a new commit.

Reply:
> We added a default implementation for since it was changing `InboundMessage`. I think we can now move away from default implementation since we have a`NativeInboundMessage`. Let me try to work this in a new commit.

But InboundMessage implements ProtocolInboundMessage, if we don't add default method, we would have breaking change for InboundMessage

Reply:
But aren't we just introducing a new method in `InboundMessage`? Would that also lead to a breaking change?

Reply:
The method is `default`, so no - the change is not breaking

Reply:
No I mean if we remove the `default` implementation from `ProtocolInboundMessage` and keep the responsibility of implementations on the implementers, we would add a new method in `InboundMessage` for `getProtocol` which would delegate to `NativeInboundMessage`. Would that be a breaking change then?